File: app/static/sdks/bluepay.py
from __future__ import annotations

import uuid
from dataclasses import dataclass
from typing import Any

import httpx


@dataclass(slots=True)
class BluePayClient:
    api_key: str
    base_url: str
    timeout: float = 15.0

    def _headers(self, idempotency_key: str | None = None) -> dict[str, str]:
        headers = {"X-API-Key": self.api_key, "Accept": "application/json"}
        if idempotency_key:
            headers["Idempotency-Key"] = idempotency_key
        return headers

    def create_invoice(self, *, amount_toman: int, order_id: str, description: str | None = None, fee_mode: str = "default", idempotency_key: str | None = None) -> dict[str, Any]:
        payload = {"amount_toman": amount_toman, "order_id": order_id, "description": description, "fee_mode": fee_mode}
        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(f"{self.base_url.rstrip('/')}/api/v1/invoices", headers=self._headers(idempotency_key or f"invoice-{order_id}"), json=payload)
        response.raise_for_status()
        return response.json()

    def get_invoice(self, payment_id: str) -> dict[str, Any]:
        with httpx.Client(timeout=self.timeout) as client:
            response = client.get(f"{self.base_url.rstrip('/')}/api/v1/invoices/{payment_id}", headers=self._headers())
        response.raise_for_status()
        return response.json()

    def create_sandbox_invoice(self, *, amount_toman: int, order_id: str, description: str | None = None) -> dict[str, Any]:
        payload = {"amount_toman": amount_toman, "order_id": order_id, "description": description}
        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(f"{self.base_url.rstrip('/')}/api/v1/sandbox/invoices", headers=self._headers(f"sandbox-{order_id}"), json=payload)
        response.raise_for_status()
        return response.json()

    def simulate_sandbox_paid(self, payment_id: str) -> dict[str, Any]:
        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(f"{self.base_url.rstrip('/')}/api/v1/sandbox/invoices/{payment_id}/simulate", headers=self._headers(), json={"result": "paid", "reference_number": f"TEST-{uuid.uuid4().hex[:12].upper()}"})
        response.raise_for_status()
        return response.json()

File: app/static/sdks/bluepay.js
export class BluePayClient {
  constructor({ apiKey, baseUrl, timeoutMs = 15000 }) {
    if (!apiKey) throw new Error("apiKey is required");
    if (!baseUrl) throw new Error("baseUrl is required");
    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.timeoutMs = timeoutMs;
  }

  async request(path, { method = "GET", body, idempotencyKey } = {}) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await fetch(`${this.baseUrl}${path}`, {
        method,
        signal: controller.signal,
        headers: {
          "X-API-Key": this.apiKey,
          "Accept": "application/json",
          "Content-Type": "application/json",
          ...(idempotencyKey ? { "Idempotency-Key": idempotencyKey } : {}),
        },
        body: body === undefined ? undefined : JSON.stringify(body),
      });
      const data = await response.json().catch(() => ({}));
      if (!response.ok) {
        throw Object.assign(new Error(data?.error?.message || data?.detail || "BluePay request failed"), {
          status: response.status,
          response,
          data,
        });
      }
      return data;
    } finally {
      clearTimeout(timer);
    }
  }

  createInvoice({ amountToman, orderId, description, feeMode = "default", callbackUrl, ttlMinutes, idempotencyKey }) {
    return this.request("/api/v1/invoices", {
      method: "POST",
      idempotencyKey: idempotencyKey || `invoice-${orderId}`,
      body: {
        amount_toman: amountToman,
        order_id: orderId,
        description,
        fee_mode: feeMode,
        ...(callbackUrl ? { callback_url: callbackUrl } : {}),
        ...(ttlMinutes !== undefined ? { ttl_minutes: ttlMinutes } : {}),
      },
    });
  }

  getInvoice(paymentId) {
    return this.request(`/api/v1/invoices/${encodeURIComponent(paymentId)}`);
  }

  getInvoiceTimeline(paymentId) {
    return this.request(`/api/v1/invoices/${encodeURIComponent(paymentId)}/timeline`);
  }

  cancelInvoice(paymentId, { idempotencyKey } = {}) {
    return this.request(`/api/v1/invoices/${encodeURIComponent(paymentId)}/cancel`, {
      method: "POST",
      idempotencyKey: idempotencyKey || `cancel-${paymentId}`,
    });
  }

  createSandboxInvoice({ amountToman, orderId, description, idempotencyKey }) {
    return this.request("/api/v1/sandbox/invoices", {
      method: "POST",
      idempotencyKey: idempotencyKey || `sandbox-${orderId}`,
      body: { amount_toman: amountToman, order_id: orderId, description },
    });
  }

  getSandboxInvoice(paymentId) {
    return this.request(`/api/v1/sandbox/invoices/${encodeURIComponent(paymentId)}`);
  }

  simulateSandbox(paymentId, { result = "paid", referenceNumber } = {}) {
    return this.request(`/api/v1/sandbox/invoices/${encodeURIComponent(paymentId)}/simulate`, {
      method: "POST",
      idempotencyKey: `sandbox-simulate-${paymentId}-${result}`,
      body: {
        result,
        ...(result === "paid" && referenceNumber ? { reference_number: referenceNumber } : {}),
      },
    });
  }
}
